lib/eslint.js

```diff
@@ -673,10 +673,17 @@ module.exports = (function() {
 
         // only do this for text
         if (text !== null) {
+            // Check and strip Unicode BOM.
+            // In JavaScript, string data is stored as UTF-16, so BOM is 0xFEFF.
+            // http://www.ecma-international.org/ecma-262/6.0/#sec-unicode-format-control-characters
+            var hasBOM = (text.charCodeAt(0) === 0xFEFF);
+            if (hasBOM) {
+                text = text.slice(1);
+            }
 
             // there's no input, just exit here
             if (text.trim().length === 0) {
-                sourceCode = new SourceCode(text, blankScriptAST);
+                sourceCode = new SourceCode(text, blankScriptAST, {hasBOM: hasBOM});
```

Comment:
I think we need to keep the text as it was originally when we pass it in here sonthe BOM is present. Then, you can calculate `hasBOM` in the constructor.